### PR TITLE
Gnocchi refactor

### PR DESCRIPTION
**Pull request by nathanielparke**
# Gnocchi Refactor
This refactor represents a comprehensive overhaul of the gnocchi codebase. 
## Purposes for the refactor 
* Migrate to spark Dataset API for increased performance 
* Increase memory efficiency through more efficient data models (namely moving creating models of reduced size for phenotypes and CalledVariants)
* Move codebase to more functional style
* Improve test coverage
## Areas of interest to a code reviewer
1. Parametrizing SparkContext and SparkSession accesses
    * I had some trouble running this current branch on the BDG cluster due to not getting the SparkSession in a correct configuration
    * `val sparkSession = SparkSession.builder().getOrCreate()` works on my local machine but in the BDG cluster there is an error relating to correctly specifying the spark master node
    * Direction on correct practices here would be appreciated
2. Parametrizing the filesystem being used
    * similar to the previous note, when running on the BDG cluster, there were issues surrounding how gnocchi accesses files and the paths that are specified 
    * How can we make the file paths more robust to a failure when there could be multiple filesystems that spark has access to?
3.  Refactor of the `org.bdgenomics.gnocchi.core.models` package
    * There is a lot of code reuse here that is limited by small differences in the traits being inherited
    * Current reason it is written this way is to accommodate the spark dataset encoders, which I could not get working with generic higher level classes
    * This meant that any operation that dealt with `Dataset` objects had to exist in a non-abstract, non-trait class
    * Are there work arounds to this? Is it worth writing tests for this package if it will be refactored into a different model in the near future (with a new spark release)? 
4. Are the maven POM's correct?
    * the maven POM's seem to work, but due to the limited understanding of our team on the power of maven, we're unsure if we are getting the most out of the tool
    * There are probably many different POM's that can build a project correctly, how do we know if ours is correct? Any pointers for Maven POM best practices?
5. Are there any obvious inefficiencies that need to be addressed?

## Work that still needs to be completed 
I am not sure if we want to address these needs before we merge this PR into the bdgenomics repo, but known future work is tracked in the list below. 
* [x] Test the `GnocchiModel` package
    * specifically, test merging of two models and the I/O of the models 
* [x] Integration tests that compare a run of plink on some input data with a run of gnocchi on the same input data
* [ ] performance testing and documentation on how to best tune the spark cluster for performance 
    * **Note:** This is very unfamiliar territory for me, but any advice on how to do this would be greatly appreciated
    * Are there inefficiencies in the way gnocchi deals with its genomic data? Inefficiencies in the `GnocchiSession` operations (Specifically the `filterVariants` and `filterSamples` functions)? 
* [x] Load in ADAM formatted variants 
    * **Note:** We currently use ADAM to load in a VCF from a specified file path, but we have not written capability to load in parquet files of ADAM VariantRDDs 
    * What is the correct use here? Should gnocchi only ingest VCF? or should it be able to take in the ADAM formatted objects?
* [ ] Model drift for QualityControlVariantModel
* [x] CLI tests
* [ ] More rigid documentation
    * adding stuff to the docs directory
    * adding method level comments
    * explaining the more complicated portions of the codebase
* [ ] More example use cases

### CODE REVIEW

**Comment by nathanielparke** [comment]
@heuermh please review when you get the chance. Thanks!

**Comment by devin-petersohn** [comment]
Jenkins, test this please.

File: examples/test.scala [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import org.bdgenomics.gnocchi.sql.GnocchiSession._
```
Comment (akmorrow13) [inline examples/test.scala]: I would rename this file so the users can run it. maybe 'run-example.scala'

File: bin/compute-gnocchi-classpath.sh [hunk 2]
```diff
@@ -1,42 +0,0 @@
-#!/usr/bin/env bash
```
Comment (akmorrow13) [inline bin/compute-gnocchi-classpath.sh:1]: Why is this file deleted?

File: README.md [hunk 3]
```diff
@@ -19,29 +19,38 @@ by setting the `MAVEN_OPTS` environment variable to `-Xmx2g -XX:MaxPermSize=1g`.
...
+echo "export PATH=[GNOCCHI INSTALLATION DIR]/gnocchi/bin:\$PATH" >> $HOME/.bashrc
+source $HOME/.bashrc
+```
+
+You can then run `gnocchi` via `gnocchi-submit`, or open a shell using `gnocchi-shell`.
+
+Test data is included. You can run with the test data by running:
+
+```
+gnocchi-submit regressPhenotypes testData/5snps10samples.vcf testData/10samples5Phenotypes2covars.txt ADDITIVE_LINEAR testData/associations -saveAsText -phenoName pheno1 -covar -covarFile testData/10samples5Phenotypes2covars.txt -covarNames pheno4,pheno5
```
Comment (devin-petersohn) [inline README.md]: This is a pretty serious command to use as an example without explanation. If you're trying to point to the test data, you could just give them the path. For the users with less experience, this will be overwhelming and may not convey what you're trying to convey.

File: README.md [hunk 4]
```diff
@@ -19,29 +19,38 @@ by setting the `MAVEN_OPTS` environment variable to `-Xmx2g -XX:MaxPermSize=1g`.
...
 ```
 
 ## Phenotype Input
 
-We accept phenotype inputs in a CSV format:
+Format phenotypes in a CSV. Both comma and tabs are accepted as delimiters.
```
Comment (devin-petersohn) [inline README.md]: `Format phenotypes in a CSV. Both comma and tabs are accepted as delimiters.` -> `Format your phenotypes files as CSV or tab-delimited text.`

File: README.md [hunk 5]
```diff
@@ -19,29 +19,38 @@ by setting the `MAVEN_OPTS` environment variable to `-Xmx2g -XX:MaxPermSize=1g`.
...
 
 ```
-Sample,Phenotype,Has Phenotype
-mySample,a phenotype,true
+SampleID    pheno1    pheno2
```
Comment (devin-petersohn) [inline README.md:49]: Is this header required? I don't see whether or not it is in the documentation.

File: bin/gnocchi-submit [hunk 6]
```diff
@@ -25,46 +27,73 @@ PRE_DD=()
...
-# Split out the CLI jar, since it will be passed to Spark as the "primary resource".
-GNOCCHI_CLI_JAR=${GNOCCHI_JARS##*,}
-GNOCCHI_JARS=$(echo "$GNOCCHI_JARS" | rev | cut -d',' -f2- | rev)
+GNOCCHI_CLI_JAR="${ASSEMBLY_DIR}/${ASSEMBLY_JARS}"
 
-# append GNOCCHI_JARS to the --jars option, if any
-SPARK_ARGS=$("$SCRIPT_DIR"/bin/append_to_option.py , --jars $GNOCCHI_JARS $SPARK_ARGS)
+# Allow main class to be overridden
+if [ -z "$GNOCCHI_MAIN" ]; then
+  GNOCCHI_MAIN="net.fnothaft.gnocchi.cli.GnocchiMain"
```
Comment (devin-petersohn) [inline bin/gnocchi-submit]: `org.bdgenomics.gnocchi.cli.GnocchiMain`

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
Many of these files have changed permissions from 644 -> 755, but they should not be executable.

I left some comments where I thought necessary, but I haven't gone through the whole thing yet. I can't tell @AmplabJenkins what to do yet, so I'm going to get the permissions fixed on it so we can get the test suite evaluated.

File: bin/compute-gnocchi-classpath.sh (hunk 2 above)
Comment (nathanielparke) [reply bin/compute-gnocchi-classpath.sh:1]: I wasn't using it, and didn't think there was a need for it. What is the typical use case of these?

**Comment by devin-petersohn** [comment]
Jenkins, test this please.

File: bin/compute-gnocchi-classpath.sh (hunk 2 above)
Comment (akmorrow13) [reply bin/compute-gnocchi-classpath.sh:1]: Generally we have used this file in the past to avoid redundancy when computing the classpath in shell and submit scripts. I would keep this file, as well as compute-gnocchi-jars, and call these in gnocchi-submit and gnocchi-shell scripts

**Comment by devin-petersohn** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by kunalgosar** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by nathanielparke** [comment]
Jenkins, test this please.

**Comment by heuermh** [comment]
Looks like the Jenkins build source code style checker is giving you trouble:
```
test -n ' M gnocchi-core/src/main/scala/org/bdgenomics/gnocchi/sql/GnocchiKryoRegistrator.scala
 M gnocchi-core/src/main/scala/org/bdgenomics/gnocchi/utils/Timers.scala
 M gnocchi-core/src/test/scala/org/bdgenomics/gnocchi/algorithms/siteregression/SiteRegressionSuite.scala'

Please run './scripts/format-source'
```

**Comment by nathanielparke** [comment]
Jenkins, test this please.
